File: backend/app/api/routes/ai.py
import json

from fastapi import APIRouter, Depends
from fastapi.responses import StreamingResponse

from app.api.deps import get_ai_service
from app.schemas.ai import AiProviderConfig, AiProviderPublicConfig, AiProviderTestRequest, AiProviderTestResponse, AiQueryRequest, AiQueryResponse
from app.services.ai_service import AiService

router = APIRouter(prefix="/ai", tags=["ai"])


@router.get("/config", response_model=AiProviderPublicConfig)
def get_ai_config(service: AiService = Depends(get_ai_service)) -> AiProviderPublicConfig:
    return service.get_config()


@router.put("/config", response_model=AiProviderPublicConfig)
def save_ai_config(config: AiProviderConfig, service: AiService = Depends(get_ai_service)) -> AiProviderPublicConfig:
    return service.save_config(config)


@router.post("/test", response_model=AiProviderTestResponse)
async def test_ai_config(request: AiProviderTestRequest, service: AiService = Depends(get_ai_service)) -> AiProviderTestResponse:
    return await service.test_connection(request)


@router.post("/query", response_model=AiQueryResponse)
async def query_ai(request: AiQueryRequest, service: AiService = Depends(get_ai_service)) -> AiQueryResponse:
    return await service.query(request)


@router.post("/stream")
async def stream_ai(request: AiQueryRequest, service: AiService = Depends(get_ai_service)) -> StreamingResponse:
    async def events():
        response = await service.query(request)
        for word in response.message.content.split(" "):
            yield f"data: {json.dumps({'type': 'content', 'content': word + ' '})}\n\n"
        for citation in response.message.citations or []:
            yield f"data: {json.dumps({'type': 'citation', 'citation': citation.model_dump(mode='json', by_alias=True)})}\n\n"
        yield f"data: {json.dumps({'type': 'done'})}\n\n"

    return StreamingResponse(events(), media_type="text/event-stream")

File: apps/frontend/src/shared/services/api/ai.ts
import { api } from './client';
import type { RequestConfig } from './client';
import type {
  AiConversationResponse,
  AiQueryRequest,
  AiQueryResponse,
  AiProviderCapabilitiesResponse,
  AiProviderConfig,
  AiProviderPublicConfig,
  AiProviderTestRequest,
  AiProviderTestResponse,
} from './types';

export const aiService = {
  getProviders(config?: RequestConfig): Promise<AiProviderCapabilitiesResponse> {
    return api.get('/ai/providers', config);
  },

  getConfig(config?: RequestConfig): Promise<AiProviderPublicConfig> {
    return api.get('/ai/config', config);
  },

  saveConfig(request: AiProviderConfig, config?: RequestConfig): Promise<AiProviderPublicConfig> {
    return api.put('/ai/config', request, config);
  },

  testConfig(request: AiProviderTestRequest, config?: RequestConfig): Promise<AiProviderTestResponse> {
    return api.post('/ai/test', request, config);
  },

  query(request: AiQueryRequest, config?: RequestConfig): Promise<AiQueryResponse> {
    return api.post('/ai/query', request, config);
  },

  listConversations(repositoryId: string, config?: RequestConfig): Promise<AiConversationResponse> {
    return api.get(`/ai/conversations?repositoryId=${encodeURIComponent(repositoryId)}`, config);
  },

};
